ArduCopter/GCS_Mavlink.cpp

```diff
@@ -92,6 +92,10 @@ MAV_STATE GCS_MAVLINK_Copter::vehicle_system_status() const
 
 void GCS_MAVLINK_Copter::send_position_target_global_int()
 {
+    if (!AP::ahrs().healthy()) {
```

Comment:
Both calls that can fail here are boolean - shouldn't this check be further down inside those instead of up here?

This might catch the same problem elsewhere.

Reply:
Placed this checked in front of `target.change_alt_frame` so that `change_alt_frame()` isn't called on an invalid Location.

Reply:
How does one reproduce this problem?

Reply:
Here is the autotest that fails without something here `autotest.py build.Copter test.Copter.RangeFinderDrivers`. The locations being gathered by `get_wp(target)` are all 0, 0, 0 for a good while and that cause the test to dump here https://github.com/ArduPilot/ardupilot/blob/bc38d8a78c2770d8dd44c438be614d89266a8995/libraries/AP_Common/Location.cpp#L108

I'm probably chasing an incorrect way to solve this. Though it made some sense to not send targets out when we don't know where we are yet.

Reply:
I think the problem here is essentially intractable given the changes you're making here.

Some of the codepaths you're crossing here are attempting to create a Location object given a Vector3f-from-origin.  If we don't have an origin that really isn't going to work - and you really can't just "promise" you'll get it right later when information is requested from the Location object (or fail at at that point) as the information's just been lost at that point.  We'd need to carry the Vector3f in the Location object to achieve the latter.

It's this check being removed in `get_wp_dest_loc` which is key:
```
-    if (!AP::ahrs().get_origin(destination)) {
-        return false;
-    }
```

... so I think the premise of making these functions `void` is itself void...

Simplifications may still be possible here - I haven't stared at it too hard :-)

Reply:
What you describe, is basically why I had the `if (!AP::ahrs().healthy())` check at the beginning of the codepaths in `send_position_target_global_int()` originally instead of at the end in `get_wp_dest_loc`. Thus to not even try to construct any Location objects before we have an origin.

There isn't a `AP::ahrs().origin_is_set()` though like there is for home.... or I would have checked for that.

Reply:
You can't have a function hand back an invalid location without it indicating failure somehow - thus the reason it is boolean now.

If this isn't a bug now, it will be in the future when someone calls this for some other reason without doing the appropriate checks.  Pretty sure these should remain boolean.

BTW, healthy() isn't actually the right check; AFAIK you really just need origin-is-set.

Reply:
Makes sense for why the check should be in the getter. Thanks for looking at it further and helping me figure it out.

Got rid of the bool changes.

PS: I did try creating a AP::ahrs().origin_is_set() since it did not exist. But it ended being a wash in size for copter and +16-40B for every other vehicle.